File: car_reminders.php
<?php
// car_reminders.php — PHP версия

$dataFile = 'reminders.json';

function loadReminders() {
    global $dataFile;
    if (file_exists($dataFile)) {
        $json = file_get_contents($dataFile);
        return json_decode($json, true) ?: [];
    }
    return [];
}

function saveReminders($reminders) {
    global $dataFile;
    file_put_contents($dataFile, json_encode($reminders, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
}

$reminders = loadReminders();

function color($text, $code) {
    return "\033[{$code}m{$text}\033[0m";
}

while (true) {
    echo "\n" . color("🔔 Дневник напоминаний (PHP)", '36') . "\n";
    echo "1. Добавить напоминание\n";
    echo "2. Показать все напоминания\n";
    echo "3. Показать предстоящие напоминания\n";
    echo "4. Поиск напоминаний\n";
    echo "5. Удалить напоминание\n";
    echo "6. Отметить как выполненное\n";
    echo "7. Статистика\n";
    echo "8. Выход\n";
    echo "Выберите действие: ";
    $choice = trim(fgets(STDIN));

    switch ($choice) {
        case '1':
            echo "Название: ";
            $title = trim(fgets(STDIN));
            echo "Дата (ГГГГ-ММ-ДД): ";
            $date = trim(fgets(STDIN));
            echo "Пробег (км): ";
            $mileage = (int) trim(fgets(STDIN));
            echo "Описание: ";
            $desc = trim(fgets(STDIN));
            echo "Приоритет (низкий/средний/высокий): ";
            $priority = strtolower(trim(fgets(STDIN)));
            if (!in_array($priority, ['низкий', 'средний', 'высокий'])) $priority = 'средний';
            $id = count($reminders) + 1;
            $reminders[] = [
                'id' => $id,
                'title' => $title,
                'date' => $date,
                'mileage' => $mileage,
                'description' => $desc,
                'priority' => $priority,
                'done' => false
            ];
            saveReminders($reminders);
            echo color("✅ Напоминание добавлено (ID: $id)\n", '32');
            break;

        case '2':
            if (empty($reminders)) {
                echo color("Нет напоминаний.\n", '33');
            } else {
                printf(color("%-4s %-20s %-12s %-10s %-10s %-12s\n", '36'), "ID", "Название", "Дата", "Пробег", "Приоритет", "Статус");
                echo str_repeat("-", 80) . "\n";
                foreach ($reminders as $r) {
                    $status = $r['done'] ? "✅ Выполнено" : "⏳ Ожидает";
                    $color = $r['priority'] == 'низкий' ? '32' : ($r['priority'] == 'средний' ? '33' : '31');
                    printf("%-4d %-20s %-12s %-10d %s%-10s\033[0m %-12s\n", $r['id'], $r['title'], $r['date'], $r['mileage'], color("", $color), $r['priority'], $status);
                }
            }
            break;

        case '3':
            echo "Количество дней (по умолч. 7): ";
            $daysStr = trim(fgets(STDIN));
            $days = $daysStr ? (int)$daysStr : 7;
            $now = new DateTime();
            $found = false;
            echo color("📅 Напоминания на ближайшие $days дней:\n", '36');
            foreach ($reminders as $r) {
                if ($r['done']) continue;
                try {
                    $d = new DateTime($r['date']);
                    $diff = $now->diff($d);
                    if ($d >= $now && $d <= (clone $now)->modify("+$days days")) {
                        echo "  {$r['id']}: {$r['title']} — {$r['date']} (осталось {$diff->days} дн)\n";
                        $found = true;
                    }
                } catch (Exception $e) {}
            }
            if (!$found) echo color("Нет напоминаний на ближайшие $days дней.\n", '32');
            break;

        case '4':
            echo "Введите ключевое слово: ";
            $keyword = strtolower(trim(fgets(STDIN)));
            $found = false;
            foreach ($reminders as $r) {
                if (stripos($r['title'], $keyword) !== false || stripos($r['description'], $keyword) !== false) {
                    echo "{$r['id']}: {$r['title']} | {$r['date']} | {$r['priority']} | " . ($r['done'] ? '✅' : '⏳') . "\n";
                    $found = true;
                }
            }
            if (!$found) echo color("Ничего не найдено.\n", '33');
            break;

        case '5':
            if (empty($reminders)) { echo color("Нет напоминаний.\n", '33'); break; }
            foreach ($reminders as $r) {
                echo "{$r['id']}: {$r['title']} — {$r['date']}\n";
            }
            echo "Введите ID для удаления: ";
            $id = (int) trim(fgets(STDIN));
            $index = array_search($id, array_column($reminders, 'id'));
            if ($index !== false) {
                array_splice($reminders, $index, 1);
                saveReminders($reminders);
                echo color("✅ Напоминание удалено.\n", '32');
            } else {
                echo color("❌ Напоминание не найдено.\n", '31');
            }
            break;

        case '6':
            if (empty($reminders)) { echo color("Нет напоминаний.\n", '33'); break; }
            foreach ($reminders as $r) {
                echo "{$r['id']}: {$r['title']} — {$r['date']} (" . ($r['done'] ? '✅ выполнено' : '⏳ ожидает') . ")\n";
            }
            echo "Введите ID для отметки как выполненное: ";
            $id = (int) trim(fgets(STDIN));
            foreach ($reminders as &$r) {
                if ($r['id'] == $id) {
                    $r['done'] = true;
                    saveReminders($reminders);
                    echo color("✅ Напоминание отмечено как выполненное.\n", '32');
                    break 2;
                }
            }
            echo color("❌ Напоминание не найдено.\n", '31');
            break;

        case '7':
            $total = count($reminders);
            $done = count(array_filter($reminders, function($r) { return $r['done']; }));
            $pending = $total - $done;
            $now = new DateTime();
            $overdue = 0;
            foreach ($reminders as $r) {
                if (!$r['done']) {
                    try {
                        if (new DateTime($r['date']) < $now) $overdue++;
                    } catch (Exception $e) {}
                }
            }
            echo color("📊 Статистика:\n", '36');
            echo "  Всего напоминаний: $total\n";
            echo "  Выполнено: $done\n";
            echo "  Ожидает: $pending\n";
            echo "  Просрочено: $overdue\n";
            break;

        case '8':
            echo "До свидания!\n";
            exit(0);

        default:
            echo color("Неверный выбор.\n", '31');
    }
}
?>

File: car_reminders.java
// car_reminders.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.time.*;
import java.time.format.*;

class Reminder {
    int id;
    String title;
    String date;
    int mileage;
    String description;
    String priority;
    boolean done;

    Reminder(int id, String title, String date, int mileage, String description, String priority, boolean done) {
        this.id = id;
        this.title = title;
        this.date = date;
        this.mileage = mileage;
        this.description = description;
        this.priority = priority;
        this.done = done;
    }

    String toJson() {
        return String.format("{\"id\":%d,\"title\":\"%s\",\"date\":\"%s\",\"mileage\":%d,\"description\":\"%s\",\"priority\":\"%s\",\"done\":%b}",
                id, title, date, mileage, description, priority, done);
    }

    static Reminder fromJson(String json) {
        // Упрощённый парсинг для демонстрации
        return null;
    }
}

public class car_reminders {
    private static final String DATA_FILE = "reminders.json";
    private static List<Reminder> reminders = new ArrayList<>();
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        load();
        while (true) {
            System.out.println("\n\u001B[36m🔔 Дневник напоминаний (Java)\u001B[0m");
            System.out.println("1. Добавить напоминание");
            System.out.println("2. Показать все напоминания");
            System.out.println("3. Показать предстоящие напоминания");
            System.out.println("4. Поиск напоминаний");
            System.out.println("5. Удалить напоминание");
            System.out.println("6. Отметить как выполненное");
            System.out.println("7. Статистика");
            System.out.println("8. Выход");
            System.out.print("Выберите действие: ");
            String choice = scanner.nextLine();
            switch (choice) {
                case "1": addReminder(); break;
                case "2": listAll(); break;
                case "3": listUpcoming(); break;
                case "4": searchReminders(); break;
                case "5": deleteReminder(); break;
                case "6": markDone(); break;
                case "7": showStats(); break;
                case "8": System.out.println("До свидания!"); return;
                default: System.out.println("\u001B[31mНеверный выбор.\u001B[0m");
            }
        }
    }

    private static void load() {
        try {
            String content = new String(Files.readAllBytes(Paths.get(DATA_FILE)));
            // Упрощённо, в реальности нужен JSON-парсер
            reminders = new ArrayList<>();
        } catch (IOException e) {
            reminders = new ArrayList<>();
        }
    }

    private static void save() {
        try {
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < reminders.size(); i++) {
                sb.append(reminders.get(i).toJson());
                if (i < reminders.size() - 1) sb.append(",");
            }
            sb.append("]");
            Files.write(Paths.get(DATA_FILE), sb.toString().getBytes());
        } catch (IOException e) {
            System.out.println("Ошибка сохранения.");
        }
    }

    private static void addReminder() {
        System.out.print("Название: ");
        String title = scanner.nextLine();
        System.out.print("Дата (ГГГГ-ММ-ДД): ");
        String date = scanner.nextLine();
        System.out.print("Пробег (км): ");
        int mileage = Integer.parseInt(scanner.nextLine());
        System.out.print("Описание: ");
        String desc = scanner.nextLine();
        System.out.print("Приоритет (низкий/средний/высокий): ");
        String priority = scanner.nextLine().toLowerCase();
        if (!priority.equals("низкий") && !priority.equals("средний") && !priority.equals("высокий")) {
            priority = "средний";
        }
        int id = reminders.size() + 1;
        reminders.add(new Reminder(id, title, date, mileage, desc, priority, false));
        save();
        System.out.println("\u001B[32m✅ Напоминание добавлено (ID: " + id + ")\u001B[0m");
    }

    private static void listAll() {
        if (reminders.isEmpty()) {
            System.out.println("\u001B[33mНет напоминаний.\u001B[0m");
            return;
        }
        System.out.printf("\u001B[36m%-4s %-20s %-12s %-10s %-10s %-12s\u001B[0m\n", "ID", "Название", "Дата", "Пробег", "Приоритет", "Статус");
        System.out.println("-".repeat(80));
        for (Reminder r : reminders) {
            String status = r.done ? "✅ Выполнено" : "⏳ Ожидает";
            String color = r.priority.equals("низкий") ? "\u001B[32m" :
                           r.priority.equals("средний") ? "\u001B[33m" : "\u001B[31m";
            System.out.printf("%-4d %-20s %-12s %-10d %s%-10s\u001B[0m %-12s\n", r.id, r.title, r.date, r.mileage, color, r.priority, status);
        }
    }

    private static void listUpcoming() {
        System.out.print("Количество дней (по умолч. 7): ");
        String daysStr = scanner.nextLine();
        int days = daysStr.isEmpty() ? 7 : Integer.parseInt(daysStr);
        LocalDate now = LocalDate.now();
        System.out.printf("\u001B[36m📅 Напоминания на ближайшие %d дней:\u001B[0m\n", days);
        for (Reminder r : reminders) {
            if (r.done) continue;
            try {
                LocalDate d = LocalDate.parse(r.date);
                if (d.isAfter(now) && d.isBefore(now.plusDays(days))) {
                    long diff = d.toEpochDay() - now.toEpochDay();
                    System.out.printf("  %d: %s — %s (осталось %d дн)\n", r.id, r.title, r.date, diff);
                }
            } catch (Exception e) {}
        }
    }

    private static void searchReminders() {
        System.out.print("Введите ключевое слово: ");
        String keyword = scanner.nextLine().toLowerCase();
        boolean found = false;
        for (Reminder r : reminders) {
            if (r.title.toLowerCase().contains(keyword) || r.description.toLowerCase().contains(keyword)) {
                System.out.printf("%d: %s | %s | %s | %s\n", r.id, r.title, r.date, r.priority, r.done ? "✅" : "⏳");
                found = true;
            }
        }
        if (!found) System.out.println("\u001B[33mНичего не найдено.\u001B[0m");
    }

    private static void deleteReminder() {
        listAll();
        System.out.print("Введите ID для удаления: ");
        int id = Integer.parseInt(scanner.nextLine());
        boolean removed = reminders.removeIf(r -> r.id == id);
        if (removed) {
            save();
            System.out.println("\u001B[32m✅ Напоминание удалено.\u001B[0m");
        } else {
            System.out.println("\u001B[31m❌ Напоминание не найдено.\u001B[0m");
        }
    }

    private static void markDone() {
        listAll();
        System.out.print("Введите ID для отметки как выполненное: ");
        int id = Integer.parseInt(scanner.nextLine());
        for (Reminder r : reminders) {
            if (r.id == id) {
                r.done = true;
                save();
                System.out.println("\u001B[32m✅ Напоминание отмечено как выполненное.\u001B[0m");
                return;
            }
        }
        System.out.println("\u001B[31m❌ Напоминание не найдено.\u001B[0m");
    }

    private static void showStats() {
        int total = reminders.size();
        int done = 0, overdue = 0;
        LocalDate now = LocalDate.now();
        for (Reminder r : reminders) {
            if (r.done) done++;
            else {
                try {
                    if (LocalDate.parse(r.date).isBefore(now)) overdue++;
                } catch (Exception e) {}
            }
        }
        System.out.println("\u001B[36m📊 Статистика:\u001B[0m");
        System.out.println("  Всего напоминаний: " + total);
        System.out.println("  Выполнено: " + done);
        System.out.println("  Ожидает: " + (total - done));
        System.out.println("  Просрочено: " + overdue);
    }
}
